BENCH-274: Add the Tag entity and endpoints

bethcryer opened the pull request:
Includes unit tests and integration tests for Tags

bethcryer committed:
Added Tag entity and endpoints

bethcryer committed:
Added unit tests and integration tests

milkshakeuk commented on src/Answer.King.Api/Controllers/TagsController.cs:
```diff
@@ -0,0 +1,239 @@
+    // GET api/tags
+    [HttpGet]
+    [ProducesResponseType(typeof(IEnumerable<Domain.Inventory.Tag>), StatusCodes.Status200OK)]
+    [SwaggerOperation(Tags = new[] { "Tags" })]
```
You don't explicitly need the Attribute here since it will auto tag them with the controller's name. i.e. `{Tags}Controller`.

Same with all the actions on this controller.

milkshakeuk commented on src/Answer.King.Api/Services/Extensions/OrderExtensions.cs:
```diff
@@ -6,7 +6,7 @@ namespace Answer.King.Api.Services.Extensions;
 {
-    public static void AddOrRemoveLineItems(this Order order, RequestModels.Order orderChanges, IList<Product> domainProducts, IList<Category> domainCategories)
+    public static void AddOrRemoveLineItems(this Order order, RequestModels.Order orderChanges, IList<Product> domainProducts, IList<Category> domainCategories, IList<Tag> domainTags)
     {
```
does the order care about tags? isn't that just an inventory thing?

bethcryer replied:
Think you might be right - I believe the categories don't need to be returned as part of an order either, so I can remove that too

milkshakeuk commented on src/Answer.King.Infrastructure/Repositories/Mappings/CategoryFactory.cs:
```diff
@@ -9,6 +9,16 @@ namespace Answer.King.Infrastructure.Repositories.Mappings;
 
 internal static class CategoryFactory
 {
+    private static ConstructorInfo? CategoryConstructor
```
don't do it this way as it will execute this code every time the propery is acccessed, it needs to be the same as described yesterday on slack i.e. 

```
private static ConstructorInfo? CategoryConstructor { get; } = typeof(Category)
            .GetConstructors(BindingFlags.Instance | BindingFlags.NonPublic)
            .SingleOrDefault(c => c.IsPrivate && c.GetParameters().Length > 0);
```

Same for the others too.

milkshakeuk reviewed (changes requested):
just a few changes.

bethcryer committed:
Removed categories and tags from the products in Order entity

Fixed tests

Fixed code smells

Changed tests again

fixed tests once more...

Changed tests to async Task functions

milkshakeuk commented on tests/Answer.King.Api.UnitTests/Controllers/CategoriesControllerTests.cs:
```diff
@@ -102,14 +103,27 @@ public void Post_Method_HasCorrectVerbAttribute()
-    [Fact(Skip = "This test needs to be written, but not sure due best way due to Domain.Inventory.Category protection level")]
-    public void Post_ValidRequestCallsGetAction_ReturnsNewCategory()
+    [Fact]
+    public async void Post_ValidRequestCallsGetAction_ReturnsNewCategory()
```
Avoid async void... its [bad practice](https://learn.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming#avoid-async-void).

Do async Task instead.

milkshakeuk replied:
async void?

milkshakeuk commented on tests/Answer.King.Api.UnitTests/Controllers/TagsControllerTests.cs:
```diff
@@ -0,0 +1,387 @@
+    }
+
+    [Fact]
+    public async void GetAll_ValidRequest_ReturnsOkObjectResult()
```
same here.

milkshakeuk commented on tests/Answer.King.Api.UnitTests/Services/CategoryServiceTests.cs:
```diff
@@ -41,6 +41,20 @@ public async void RetireCategory_CategoryContainsProducts_ThrowsException()
     }
 
+    [Fact]
+    public async void RetireCategory_AlreadyRetired_ThrowsException()
```
same here.

milkshakeuk commented on src/Answer.King.Domain/Orders/Order.cs:
```diff
@@ -45,7 +48,7 @@ public Order()
     public IReadOnlyCollection<LineItem> LineItems => (this._LineItems as List<LineItem>)!;
 
-    public void AddLineItem(long productId, string productName, string productDescription, double price, IList<Category> productCategories, int quantity = 1)
+    public void AddLineItem(long productId, string productName, string productDescription, double price, int quantity = 1)
```
this might come back to bite us, what does the API contract say? even if we remove it from here technically, we can still hydrate it in the service but it won't be a point in time snapshot for the Category & Tags not sure if that's a bad thing or not.

milkshakeuk reviewed (changes requested):
Few more comments.

bethcryer committed:
Json converter and Factory tests

milkshakeuk committed:
Refactored Tests so tests are in correct (new) project.

milkshakeuk committed:
Refactored Tests so tests are in correct (new) project.

bethcryer committed:
Factories no longer static, updated project accordingly

bethcryer committed:
Reverted json converters and fixed Verify snapshots

bethcryer committed:
Added tests for data seeders

missed code smell

bethcryer committed:
Added Product tests

DWatson1993 commented on src/Answer.King.Api/Controllers/TagsController.cs:
```diff
@@ -0,0 +1,231 @@
+
+namespace Answer.King.Api.Controllers;
+
+[Route("api/[controller]")]
```
Does this need to be [Route("api/tags")] ?
I know it's the same with all the other Controller files but was something I was picked up on during the Academy

bethcryer replied:
Either way is fine really, I think we'll leave it as is unless there's a compelling reason to change

DWatson1993 commented on src/Answer.King.Api/Controllers/TagsController.cs:
```diff
@@ -0,0 +1,231 @@
+    /// <summary>
+    /// Get all tags.
+    /// </summary>
+    /// <returns></returns>
```
Could the <returns></returns> be removed as they aren't being used?

DWatson1993 commented on src/Answer.King.Api/Controllers/TagsController.cs:
```diff
@@ -0,0 +1,231 @@
+    /// Get a single tag.
+    /// </summary>
+    /// <param name="id"></param>
+    /// <returns></returns>
```
Could this be removed?

DWatson1993 commented on src/Answer.King.Api/Services/ITagService.cs:
```diff
@@ -0,0 +1,14 @@
+
+public interface ITagService
+{
+    Task<Tag> CreateTag(RequestModels.Tag createTag);
```
Would this benefit from having summaries added to them or is the controller file summaries enough to cover this?

bethcryer replied:
Don't think these need summaries no, those are just for generating the swagger documentation on the controllers

milkshakeuk replied:
sometimes it's useful to add xml docs but in this case I'm not sure., since it's quite simple and easy to tell what it does.

If you needed to generate some documentation then it might be a good idea to add some basic info here.

DWatson1993 commented on src/Answer.King.Domain/Repositories/ITagRepository.cs:
```diff
@@ -0,0 +1,7 @@
+namespace Answer.King.Domain.Repositories;
+
+public interface ITagRepository : IAggregateRepository<Tag>
+{
```
Is there something missing from here?

bethcryer replied:
Nah this is just implementing IAggregateRepository<Tag> - I suppose we could do that directly in TagRepository, but having its own interface just makes things a bit clearer for reading

milkshakeuk commented on src/Answer.King.Domain/Repositories/Models/Product.cs:
```diff
@@ -52,6 +56,10 @@ public Product(string name, string description, double price)
 
     public IReadOnlyCollection<CategoryId> Categories => this._Categories;
 
+    private HashSet<TagId> _Tags { get; }
```
shouldn't this be `_tags` ?

milkshakeuk commented on tests/Answer.King.Api.UnitTests/Controllers/CategoriesControllerTests.cs:
```diff
@@ -137,6 +151,27 @@ public async void Put_NullCategory_ReturnsNotFoundResult()
     }
 
+    [Fact]
+    public async void Put_ProductIdNotValid_ReturnsValidationProblem()
```
async void.

bethcryer committed:
Removed async voids and fixed remaining iserializable exceptions

dsdf

DWatson1993 commented on tests/Answer.King.Api.UnitTests/Common/CustomConverters/CategoryIdJsonConverterTests.cs:
```diff
@@ -0,0 +1,51 @@
+        var jsonReader = new Utf8JsonReader(jsonUtf8Bytes);
+        jsonReader.Read();
+
+        var tagIdJsonConverter = new CategoryIdJsonConverter();
```
Is this the correct variable name?

DWatson1993 commented on tests/Answer.King.Api.UnitTests/Common/CustomConverters/ProductIdJsonConverterTests.cs:
```diff
@@ -0,0 +1,51 @@
+        var jsonReader = new Utf8JsonReader(jsonUtf8Bytes);
+        jsonReader.Read();
+
+        var tagIdJsonConverter = new ProductIdJsonConverter();
```
variable name correct?

DWatson1993 commented on tests/Answer.King.Api.UnitTests/Common/CustomConverters/TagIdJsonConverterTests.cs:
```diff
@@ -0,0 +1,51 @@
+        var jsonReader = new Utf8JsonReader(jsonUtf8Bytes);
+        jsonReader.Read();
+
+        var tagIdJsonConverter = new TagIdJsonConverter();
```
variable name correct?

bethcryer replied:
this one is :P

milkshakeuk commented on src/Answer.King.Domain/Repositories/Models/Product.cs:
```diff
@@ -60,18 +68,31 @@ public void AddCategory(CategoryId category)
         }
-
-        this._Categories.Add(category);
+        this._categories.Add(category);
```
shouldn't there be a new line between the brace and this line?

bethcryer committed:
Removed unused documentation comments + fixed inaccurate variable names

DWatson1993 commented on tests/Answer.King.Api.UnitTests/Common/CustomConverters/CategoryIdJsonConverterTests.cs:
```diff
@@ -0,0 +1,51 @@
+        // Act
+        var result = categoryIdJsonConverter.Read(ref jsonReader, typeof(long), new JsonSerializerOptions());
+
+        // Act / Assert
```
Needs changing to // Assert

DWatson1993 commented on tests/Answer.King.Api.UnitTests/Common/CustomConverters/ProductIdJsonConverterTests.cs:
```diff
@@ -0,0 +1,51 @@
+        // Act
+        var result = productIdJsonConverter.Read(ref jsonReader, typeof(long), new JsonSerializerOptions());
+
+        // Act / Assert
```
Just assert

bethcryer committed:
Fixed factory tests

DWatson1993 commented on tests/Answer.King.Infrastructure.UnitTests/Repositories/Factories/PaymentFactoryTests.cs:
```diff
@@ -0,0 +1,47 @@
+[TestCategory(TestType.Unit)]
+public class PaymentFactoryTests
+{
+    private static readonly PaymentFactory paymentFactory = new();
```
Does this need to be caps like the others? I think in Product and Tags tests

milkshakeuk replied:
these will change with the next PR which brings in style cop instead of rolsynator

DWatson1993 commented on tests/Answer.King.Infrastructure.UnitTests/Repositories/Factories/ProductFactoryTests.cs:
```diff
@@ -0,0 +1,47 @@
+[TestCategory(TestType.Unit)]
+public class ProductFactoryTests
+{
+    private static readonly ProductFactory productFactory = new();
```
Caps like the others?

milkshakeuk replied:
these will change with the next PR which brings in style cop instead of rolsynator

DWatson1993 commented on tests/Answer.King.Infrastructure.UnitTests/Repositories/Factories/TagFactoryTests.cs:
```diff
@@ -0,0 +1,51 @@
+[TestCategory(TestType.Unit)]
+public class TagFactoryTests
+{
+    private static readonly TagFactory tagFactory = new();
```
Caps like the others?

milkshakeuk replied:
these will change with the next PR which brings in style cop instead of rolsynator
